File: src/main/java/de/bixilon/minosoft/protocol/protocol/InByteBuffer.java
/*
 * Minosoft
 * Copyright (C) 2020 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program.If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */

package de.bixilon.minosoft.protocol.protocol;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import de.bixilon.minosoft.data.Directions;
import de.bixilon.minosoft.data.commands.CommandArgumentNode;
import de.bixilon.minosoft.data.commands.CommandLiteralNode;
import de.bixilon.minosoft.data.commands.CommandNode;
import de.bixilon.minosoft.data.commands.CommandRootNode;
import de.bixilon.minosoft.data.entities.Poses;
import de.bixilon.minosoft.data.mappings.LegacyResourceLocation;
import de.bixilon.minosoft.data.mappings.ResourceLocation;
import de.bixilon.minosoft.data.text.ChatComponent;
import de.bixilon.minosoft.protocol.network.connection.Connection;
import de.bixilon.minosoft.util.Util;
import de.bixilon.minosoft.util.nbt.tag.*;
import glm_.vec2.Vec2i;
import glm_.vec3.Vec3;
import glm_.vec3.Vec3i;
import org.checkerframework.common.value.qual.IntRange;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.UUID;

public class InByteBuffer {
    private final Connection connection;
    private final byte[] bytes;
    private int position;

    public InByteBuffer(byte[] bytes, Connection connection) {
        this.bytes = bytes;
        this.connection = connection;
    }

    public InByteBuffer(InByteBuffer buffer) {
        this.bytes = buffer.getBytes();
        this.position = buffer.getPosition();
        this.connection = buffer.getConnection();
    }

    public short readShort() {
        return (short) (((readUnsignedByte()) << 8) | (readUnsignedByte()));
    }

    @IntRange(from = 0, to = ((int) Short.MAX_VALUE) * 2)
    public int readUnsignedShort() {
        return readShort() & 0xFFFF;
    }

    public int readInt() {
        return ((readUnsignedByte() << 24) | (readUnsignedByte() << 16) | (readUnsignedByte() << 8) | (readUnsignedByte()));
    }

    public byte[] readBytes(int count) {
        if (count > ProtocolDefinition.PROTOCOL_PACKET_MAX_SIZE) {
            throw new IllegalArgumentException("Trying to allocate to much memory");
        }
        byte[] ret = new byte[count];
        System.arraycopy(this.bytes, this.position, ret, 0, count);
        this.position += count;
        return ret;
    }

    public long readLong() {
        return (((long) readUnsignedByte() << 56) | ((long) readUnsignedByte() << 48) | ((long) readUnsignedByte() << 40) | ((long) readUnsignedByte() << 32) | ((long) readUnsignedByte() << 24) | (readUnsignedByte() << 16) | (readUnsignedByte() << 8) | (readUnsignedByte()));
    }

    public double readFixedPointNumberInt() {
        return readInt() / 32.0D;
    }

    public String readString() {
        String string = new String(readBytes(readVarInt()), StandardCharsets.UTF_8);
        if (string.length() > ProtocolDefinition.STRING_MAX_LEN) {
            throw new IllegalArgumentException(String.format("String max string length exceeded %d > %d", string.length(), ProtocolDefinition.STRING_MAX_LEN));
        }
        return string;
    }

    public long readVarLong() {
        int byteCount = 0;
        long result = 0;
        byte read;
        do {
            read = readByte();
            result |= (long) (read & 0x7F) << (7 * byteCount);
            byteCount++;
            if (byteCount > 10) {
                throw new IllegalArgumentException("VarLong is too big");
            }
        } while ((read & 0x80) != 0);
        return result;
    }

    public boolean readBoolean() {
        return readByte() == 1;
    }

    public int[] readUnsignedShortsLE(int count) {
        if (count > ProtocolDefinition.PROTOCOL_PACKET_MAX_SIZE) {
            throw new IllegalArgumentException("Trying to allocate to much memory");
        }
        int[] ret = new int[count];
        for (int i = 0; i < ret.length; i++) {
            ret[i] = ((readUnsignedByte()) | (readUnsignedByte() << 8));
        }
        return ret;
    }

    public String[] readStringArray(int length) {
        if (length > ProtocolDefinition.PROTOCOL_PACKET_MAX_SIZE) {
            throw new IllegalArgumentException("Trying to allocate to much memory");
        }
        String[] ret = new String[length];
        for (int i = 0; i < length; i++) {
            ret[i] = readString();
        }
        return ret;
    }

    public String[] readStringArray() {
        return readStringArray(readVarInt());
    }

    public String readString(int length) {
        return new String(readBytes(length));
    }

    public UUID readUUID() {
        return new UUID(readLong(), readLong());
    }

    public int readVarInt() {
        int byteCount = 0;
        int result = 0;
        byte read;
        do {
            read = readByte();
            result |= (read & 0x7F) << (7 * byteCount);
            byteCount++;
            if (byteCount > 5) {
                throw new IllegalArgumentException("VarInt is too big");
            }
        } while ((read & 0x80) != 0);

        return result;
    }

    public double readFixedPointNumberByte() {
        return readByte() / 32.0D;
    }

    public JsonObject readJSON() {
        return JsonParser.parseString(readString()).getAsJsonObject();
    }

    public byte readByte() {
        return this.bytes[this.position++];
    }

    @IntRange(from = 0, to = ((int) Byte.MAX_VALUE) * 2 + 1)
    public int readUnsignedByte() {
        return (this.bytes[this.position++] & 0xFF);
    }

    public ChatComponent readChatComponent() {
        return ChatComponent.Companion.valueOf(null, null, readString());
    }

    @IntRange(from = 0)
    public int getLength() {
        return this.bytes.length;
    }

    public Directions readDirection() {
        return Directions.byId(readVarInt());
    }

    public Poses readPose() {
        return Poses.byId(readVarInt());
    }

    public NBTTag readNBT(boolean compressed) {
        if (compressed) {
            int length = readShort();
            if (length == -1) {
                // no nbt data here...
                return new CompoundTag();
            }
            try {
                return new InByteBuffer(Util.decompressGzip(readBytes(length)), this.connection).readNBT();
            } catch (IOException e) {
                // oh no
                e.printStackTrace();
                throw new IllegalArgumentException("Bad nbt");
            }
        }
        TagTypes type = TagTypes.byId(readUnsignedByte());
        if (type == TagTypes.COMPOUND) {
            // shouldn't be a subtag
            return new CompoundTag(false, this);
        }
        return readNBT(type);
    }

    public NBTTag readNBT(TagTypes tagType) {
        return switch (tagType) {
            case END -> null;
            case BYTE -> new ByteTag(this);
            case SHORT -> new ShortTag(this);
            case INT -> new IntTag(this);
            case LONG -> new LongTag(this);
            case FLOAT -> new FloatTag(this);
            case DOUBLE -> new DoubleTag(this);
            case BYTE_ARRAY -> new ByteArrayTag(this);
            case STRING -> new StringTag(this);
            case LIST -> new ListTag(this);
            case COMPOUND -> new CompoundTag(true, this);
            case INT_ARRAY -> new IntArrayTag(this);
            case LONG_ARRAY -> new LongArrayTag(this);
        };
    }

    public NBTTag readNBT() {
        return readNBT(false);
    }

    public String getBase64() {
        return getBase64(getPosition(), getBytesLeft());
    }

    public String getBase64(int pos, int length) {
        return new String(Base64.getEncoder().encode(readBytes(pos, length)));
    }

    @IntRange(from = 0, to = ProtocolDefinition.PROTOCOL_PACKET_MAX_SIZE)
    public int getPosition() {
        return this.position;
    }

    public void setPosition(int pos) {
        this.position = pos;
    }

    @IntRange(from = 0)
    public int getBytesLeft() {
        return this.bytes.length - this.position;
    }

    byte[] readBytes(int pos, int count) {
        if (count > ProtocolDefinition.PROTOCOL_PACKET_MAX_SIZE) {
            throw new IllegalArgumentException("Trying to allocate to much memory");
        }
        byte[] ret = new byte[count];
        System.arraycopy(this.bytes, pos, ret, 0, count);
        return ret;
    }

    public short readAngle() {
        return (short) (readByte() * ProtocolDefinition.ANGLE_CALCULATION_CONSTANT);
    }

    public Vec3 readEntityPosition() {
        return new Vec3(readDouble(), readDouble(), readDouble());
    }

    public double readDouble() {
        return Double.longBitsToDouble(readLong());
    }

    public Vec3 readFloatPosition() {
        return new Vec3(readFloat(), readFloat(), readFloat());
    }

    public float readFloat() {
        return Float.intBitsToFloat(readInt());
    }

    public Vec3i readBlockPositionByte() {
        return new Vec3i(readInt(), readUnsignedByte(), readInt());
    }

    public Vec3i readBlockPositionShort() {
        return new Vec3i(readInt(), readShort(), readInt());
    }

    public Vec3i readBlockPositionInteger() {
        return new Vec3i(readInt(), readInt(), readInt());
    }

    public byte[] readBytesLeft() {
        return readBytes(getBytesLeft());
    }

    public int[] readIntArray(int length) {
        if (length > ProtocolDefinition.PROTOCOL_PACKET_MAX_SIZE) {
            throw new IllegalArgumentException("Trying to allocate to much memory");
        }
        int[] ret = new int[length];
        for (int i = 0; i < length; i++) {
            ret[i] = readInt();
        }
        return ret;
    }


    public long[] readLongArray(int length) {
        if (length > ProtocolDefinition.PROTOCOL_PACKET_MAX_SIZE) {
            throw new IllegalArgumentException("Trying to allocate to much memory");
        }
        long[] ret = new long[length];
        for (int i = 0; i < length; i++) {
            ret[i] = readLong();
        }
        return ret;
    }

    public long[] readLongArray() {
        return readLongArray(readVarInt());
    }

    @Override
    public String toString() {
        return "dataLen: " + this.bytes.length + "; position: " + this.position;
    }

    public byte[] getBytes() {
        return this.bytes;
    }

    public int[] readVarIntArray(int length) {
        if (length > ProtocolDefinition.PROTOCOL_PACKET_MAX_SIZE) {
            throw new IllegalArgumentException("Trying to allocate to much memory");
        }
        int[] ret = new int[length];
        for (int i = 0; i < length; i++) {
            ret[i] = readVarInt();
        }
        return ret;
    }

    public int[] readVarIntArray() {
        return readVarIntArray(readVarInt());
    }


    public Connection getConnection() {
        return this.connection;
    }


    public CommandNode[] readCommandNodesArray() {
        int count = readVarInt();

        if (count > ProtocolDefinition.PROTOCOL_PACKET_MAX_SIZE) {
            throw new IllegalArgumentException("Trying to allocate to much memory");
        }
        CommandNode[] nodes = new CommandNode[count];
        for (int i = 0; i < nodes.length; i++) {
            nodes[i] = readCommandNode();
        }
        // resole ids
        for (CommandNode node : nodes) {
            // redirect
            if (node.getRedirectNodeId() != -1) {
                node.setRedirectNode(nodes[node.getRedirectNodeId()]);
            }

            // children
            for (int id : node.getChildrenIds()) {
                CommandNode targetNode = nodes[id];
                if (targetNode instanceof CommandArgumentNode argumentNode) {
                    node.getArgumentsChildren().add(argumentNode);
                } else if (targetNode instanceof CommandLiteralNode literalNode) {
                    node.getLiteralChildren().put(literalNode.getName(), literalNode);
                }
            }
            node.resetChildrenIds();

        }
        return nodes;
    }

    private CommandNode readCommandNode() {
        byte flags = readByte();
        return switch (CommandNode.NodeTypes.byId(flags & 0x03)) {
            case ROOT -> new CommandRootNode(flags, this);
            case LITERAL -> new CommandLiteralNode(flags, this);
            case ARGUMENT -> new CommandArgumentNode(flags, this);
        };
    }

    public ResourceLocation readResourceLocation() {
        String resourceLocation = readString();

        if (Util.doesStringContainsUppercaseLetters(resourceLocation)) {
            // just a string but wrapped into a resourceLocation (like old plugin channels MC|BRAND or ...)
            return new LegacyResourceLocation(resourceLocation);
        }
        return new ResourceLocation(resourceLocation);
    }

    public Vec2i readChunkPosition() {
        return new Vec2i(readInt(), readInt());
    }

    public long[] readVarLongArray(int size) {
        var ret = new long[size];
        for (int i = 0; i < size; i++) {
            ret[i] = readVarLong();
        }
        return ret;
    }

    public long[] readVarLongArray() {
        return readVarLongArray(readVarInt());
    }
}

File: src/main/java/de/bixilon/minosoft/protocol/protocol/InByteBuffer.kt
/*
 * Minosoft
 * Copyright (C) 2020 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */
package de.bixilon.minosoft.protocol.protocol

import de.bixilon.minosoft.data.commands.CommandArgumentNode
import de.bixilon.minosoft.data.commands.CommandLiteralNode
import de.bixilon.minosoft.data.commands.CommandNode
import de.bixilon.minosoft.data.commands.CommandRootNode
import de.bixilon.minosoft.data.direction.Directions
import de.bixilon.minosoft.data.entities.Poses
import de.bixilon.minosoft.data.registries.ResourceLocation
import de.bixilon.minosoft.data.tags.Tag
import de.bixilon.minosoft.data.text.ChatComponent
import de.bixilon.minosoft.protocol.network.connection.Connection
import de.bixilon.minosoft.util.Util
import de.bixilon.minosoft.util.json.JSONSerializer
import de.bixilon.minosoft.util.nbt.tag.NBTTagTypes
import glm_.vec2.Vec2i
import glm_.vec3.Vec3
import glm_.vec3.Vec3d
import glm_.vec3.Vec3i
import java.nio.charset.StandardCharsets
import java.util.*


open class InByteBuffer {
    private val connection: Connection?
    private val bytes: ByteArray
    var pointer = 0

    constructor(bytes: ByteArray, connection: Connection) {
        this.bytes = bytes
        this.connection = connection
    }

    constructor(buffer: InByteBuffer) {
        connection = buffer.connection
        bytes = buffer.bytes.clone()
        pointer = buffer.pointer
    }

    val size: Int
        get() = bytes.size

    val bytesLeft: Int
        get() = size - pointer

    inline fun <reified T> readArray(length: Int = readVarInt(), reader: () -> T): Array<T> {
        check(length <= size) { "Trying to allocate too much memory!" }
        val array: MutableList<T> = mutableListOf()
        for (i in 0 until length) {
            array.add(i, reader())
        }
        return array.toTypedArray()
    }

    fun readByte(): Byte {
        return bytes[pointer++]
    }

    open fun readByteArray(length: Int = readVarInt()): ByteArray {
        check(length <= bytes.size) { "Trying to allocate too much memory!" }
        val array = ByteArray(length)
        System.arraycopy(bytes, pointer, array, 0, length)
        pointer += length
        return array
    }

    fun readUnsignedByte(): Int {
        return readByte().toInt() and ((1 shl Byte.SIZE_BITS) - 1)
    }


    fun readShort(): Short {
        return (readUnsignedByte() shl Byte.SIZE_BITS or readUnsignedByte()).toShort()
    }

    fun readShortArray(length: Int = readVarInt()): ShortArray {
        check(length <= bytes.size / Short.SIZE_BYTES) { "Trying to allocate too much memory!" }
        val array = ShortArray(length)
        for (i in 0 until length) {
            array[i] = readShort()
        }
        return array
    }

    fun readUnsignedShort(): Int {
        return readShort().toInt() and ((1 shl Short.SIZE_BITS) - 1)
    }

    fun readVelocity(): Vec3d {
        return Vec3d(readShort(), readShort(), readShort()) / ProtocolDefinition.VELOCITY_CONSTANT
    }

    fun readInt(): Int {
        return (readUnsignedShort() shl Short.SIZE_BITS or readUnsignedShort())
    }

    fun readIntArray(length: Int = readVarInt()): IntArray {
        check(length <= bytes.size / Int.SIZE_BYTES) { "Trying to allocate too much memory!" }
        val array = IntArray(length)
        for (i in 0 until length) {
            array[i] = readInt()
        }
        return array
    }

    fun readUnsignedInt(): Long {
        return readInt().toLong() and ((1L shl Int.SIZE_BITS) - 1)
    }


    fun readVarInt(): Int {
        var byteCount = 0
        var result = 0
        var read: Int
        do {
            read = readUnsignedByte()
            result = result or (read and 0x7F shl (Byte.SIZE_BITS - 1) * byteCount)
            byteCount++
            require(byteCount <= Int.SIZE_BYTES + 1) { "VarInt is too big" }
        } while (read and 0x80 != 0)

        return result
    }


    @JvmOverloads
    fun readVarIntArray(length: Int = readVarInt()): IntArray {
        check(length <= bytes.size) { "Trying to allocate too much memory!" }
        val array = IntArray(length)
        for (i in 0 until length) {
            array[i] = readVarInt()
        }
        return array
    }

    fun readUnsignedVarInt(): Long {
        return readVarInt().toLong() and ((1 shl Int.SIZE_BITS) - 1).toLong()
    }


    fun readLong(): Long {
        return (readUnsignedInt() shl Int.SIZE_BITS or readUnsignedInt())
    }

    fun readLongArray(length: Int = readVarInt()): LongArray {
        check(length <= bytes.size / Long.SIZE_BYTES) { "Trying to allocate too much memory!" }
        val array = LongArray(length)
        for (i in 0 until length) {
            array[i] = readLong()
        }
        return array
    }


    fun readVarLong(): Long {
        var byteCount = 0
        var result = 0L
        var read: Int
        do {
            read = readUnsignedByte()
            result = result or ((read and 0x7F shl (Byte.SIZE_BITS - 1) * byteCount).toLong())
            byteCount++
            require(byteCount <= Long.SIZE_BYTES + 1) { "VarLong is too big" }
        } while (read and 0x80 != 0)

        return result
    }

    fun readVarLongArray(length: Int = readVarInt()): LongArray {
        check(length <= bytes.size) { "Trying to allocate too much memory!" }
        val array = LongArray(length)
        for (i in 0 until length) {
            array[i] = readVarLong()
        }
        return array
    }

    fun readFloat(): Float {
        return Float.fromBits(readInt())
    }

    fun readFloatArray(length: Int = readVarInt()): FloatArray {
        check(length <= bytes.size / Float.SIZE_BYTES) { "Trying to allocate too much memory!" }
        val array = FloatArray(length)
        for (i in 0 until length) {
            array[i] = readFloat()
        }
        return array
    }


    fun readDouble(): Double {
        return Double.fromBits(readLong())
    }

    fun readDoubleArray(length: Int = readVarInt()): DoubleArray {
        check(length <= bytes.size / Double.SIZE_BYTES) { "Trying to allocate too much memory!" }
        val array = DoubleArray(length)
        for (i in 0 until length) {
            array[i] = readDouble()
        }
        return array
    }


    fun readFixedPointNumberInt(): Double {
        return readInt() / 32.0
    }

    fun readFixedPointNumberByte(): Double {
        return readByte() / 32.0
    }

    fun readBoolean(): Boolean {
        return readUnsignedByte() == 1
    }

    @JvmOverloads
    fun readString(length: Int = readVarInt()): String {
        val string = String(readByteArray(length), StandardCharsets.UTF_8)
        check(string.length <= ProtocolDefinition.STRING_MAX_LENGTH) { "String max string length exceeded ${string.length} > ${ProtocolDefinition.STRING_MAX_LENGTH}" }
        return string
    }

    fun readStringArray(length: Int = readVarInt()): Array<String> {
        return readArray(length) { readString() }
    }

    fun readUUID(): UUID {
        return UUID(readLong(), readLong())
    }

    fun readUUIDArray(length: Int = readVarInt()): Array<UUID> {
        return readArray(length) { readUUID() }
    }

    fun readUUIDString(): UUID {
        return Util.getUUIDFromString(readString())
    }

    fun readUUIDStringArray(length: Int = readVarInt()): Array<UUID> {
        return readArray(length) { readUUIDString() }
    }


    fun readJson(): Map<String, Any> {
        return JSONSerializer.MUTABLE_MAP_ADAPTER.fromJson(readString())!!
    }

    fun readJsonArray(length: Int = readVarInt()): Array<Map<String, Any>> {
        return readArray(length) { readJson() }
    }

    open fun readChatComponent(): ChatComponent {
        return ChatComponent.of(readString(), restrictedMode = true)
    }

    fun readChatComponentArray(length: Int = readVarInt()): Array<ChatComponent> {
        return readArray(length) { readChatComponent() }
    }

    fun readDirection(): Directions {
        return Directions.VALUES[readVarInt()]
    }

    fun readPose(): Poses {
        return Poses[readVarInt()]
    }

    fun readRest(): ByteArray {
        return readByteArray(length = size - pointer)
    }

    fun readAngle(): Int {
        return (readByte() * ProtocolDefinition.ANGLE_CALCULATION_CONSTANT).toInt()
    }

    fun readVec3d(): Vec3d {
        return Vec3d(readDouble(), readDouble(), readDouble())
    }

    fun readVec3f(): Vec3 {
        return Vec3(readFloat(), readFloat(), readFloat())
    }

    fun readByteBlockPosition(): Vec3i {
        return Vec3i(readInt(), readByte(), readInt())
    }

    fun readShortBlockPosition(): Vec3i {
        return Vec3i(readInt(), readShort(), readInt())
    }

    fun readIntBlockPosition(): Vec3i {
        return Vec3i(readInt(), readInt(), readInt())
    }

    fun readResourceLocation(): ResourceLocation {
        return ResourceLocation.getResourceLocation(readString())
    }


    fun readCommandNode(): CommandNode {
        val flags = readByte().toInt()
        return when (CommandNode.NodeTypes.byId(flags and 0x03)!!) {
            CommandNode.NodeTypes.ROOT -> CommandRootNode(flags, this)
            CommandNode.NodeTypes.LITERAL -> CommandLiteralNode(flags, this)
            CommandNode.NodeTypes.ARGUMENT -> CommandArgumentNode(flags, this)
        }
    }

    @JvmOverloads
    fun readCommandNodeArray(length: Int = readVarInt()): Array<CommandNode> {
        val nodes = readArray(length) { readCommandNode() }
        for (node in nodes) {
            if (node.redirectNodeId != -1) {
                node.redirectNode = nodes[node.redirectNodeId]
            }

            for (childId in node.childrenIds) {
                when (val child = nodes[childId]) {
                    is CommandArgumentNode -> {
                        node.argumentsChildren.add(child)
                    }
                    is CommandLiteralNode -> {
                        node.literalChildren[child.name] = child
                    }
                }
            }
        }

        return nodes
    }

    fun readChunkPosition(): Vec2i {
        return Vec2i(readInt(), readInt())
    }

    fun readUnsignedShortsLE(length: Int): IntArray {
        require(length <= size) { "Trying to allocate to much memory" }
        val ret = IntArray(length)
        for (i in 0 until length) {
            ret[i] = readUnsignedByte() or (readUnsignedByte() shl 8)
        }
        return ret
    }

    fun readNBTTag(tagType: NBTTagTypes): Any? {
        return when (tagType) {
            NBTTagTypes.END -> null
            NBTTagTypes.BYTE -> readByte()
            NBTTagTypes.SHORT -> readShort()
            NBTTagTypes.INT -> readInt()
            NBTTagTypes.LONG -> readLong()
            NBTTagTypes.FLOAT -> readFloat()
            NBTTagTypes.DOUBLE -> readDouble()
            NBTTagTypes.BYTE_ARRAY -> readByteArray(readInt())
            NBTTagTypes.STRING -> readString(readUnsignedShort())
            NBTTagTypes.LIST -> {
                val listType = NBTTagTypes[readUnsignedByte()]
                val length = readInt()
                val out: MutableList<Any> = mutableListOf()
                for (i in 0 until length) {
                    readNBTTag(listType)?.let { out.add(it) }
                }
                out
            }
            NBTTagTypes.COMPOUND -> {
                val out: MutableMap<String, Any> = mutableMapOf()
                while (true) {
                    val compoundTagType = NBTTagTypes[readUnsignedByte()]
                    if (compoundTagType === NBTTagTypes.END) {
                        // end tag
                        break
                    }
                    val tagName: String = readString(readUnsignedShort())
                    val tag = readNBTTag(compoundTagType) ?: continue
                    out[tagName] = tag
                }
                out
            }
            NBTTagTypes.INT_ARRAY -> readIntArray(readInt())
            NBTTagTypes.LONG_ARRAY -> readLongArray(readInt())
        }
    }

    fun readNBTTag(compressed: Boolean): Any? {
        if (compressed) {
            val length = readShort().toInt()
            return if (length == -1) {
                // no nbt data here...
                null
            } else {
                InByteBuffer(Util.decompressGzip(readByteArray(length)), connection!!).readNBTTag(false)
            }
        }
        val type = NBTTagTypes[readUnsignedByte()]
        if (type === NBTTagTypes.COMPOUND) {
            var name = readString(readUnsignedShort()) // ToDo
        }
        return readNBTTag(type)
    }

    fun getBase64(): String {
        return String(Base64.getEncoder().encode(readRest()))
    }

    fun <T> readTag(idResolver: (Int) -> T): Pair<ResourceLocation, Tag<T>> {
        val resourceLocation = readResourceLocation()
        val ids = readVarIntArray()
        val items: MutableSet<T> = mutableSetOf()
        for (id in ids) {
            items += idResolver(id)
        }
        return Pair(resourceLocation, Tag(items.toSet()))
    }

    fun <T> readTagArray(length: Int = readVarInt(), idResolver: (Int) -> T): Map<ResourceLocation, Tag<T>> {
        return mapOf(*(readArray(length) { readTag(idResolver) }))
    }

    fun <T> readOptional(reader: () -> T): T? {
        return if (readBoolean()) {
            reader()
        } else {
            null
        }
    }
}
